### PR TITLE
fix: fix script replication

### LINKED ISSUE #2519: Eval on a single shard does not log exec to journal
When running eval on a single shard we have an optimisation for a single hop
inside Service::EvalInternal  if (CanRunSingleShardMulti(sid, *params, *tx)
When this optimisation is taking place we log to journal all the executed commands in the script with MULTI opcode but when finishing running the script we do not journal EXEC opcode.
This bug is causing replica to get stuck on execution and memory continues to grow untill oom occurs.

### PR DESCRIPTION
**Pull request by dranikpg**
Fixes #2519

### CODE REVIEW

File: src/server/transaction.cc [hunk 1]
```diff
@@ -966,6 +968,16 @@ const absl::flat_hash_set<std::string_view>& Transaction::GetMultiKeys() const {
   return multi_->frozen_keys_set;
 }
 
+void Transaction::FIX_ConcludeJournalExec() {
+  if (!multi_->shard_journal_write.front())
+    return;
+
+  if (auto journal = EngineShard::tlocal()->journal(); journal != nullptr) {
+    journal->RecordEntry(txid_, journal::Op::EXEC, db_index_, 1,
+                         unique_slot_checker_.GetUniqueSlotId(), {}, true);
```
Comment (adiholden) [inline src/server/transaction.cc:977]: the last param means that you can preempt when this is executed. I think its safer if we do not preempt here as this function runs inside a callback on the tx queue.
Now when reviewing the flow I see we can preempt when running a squeshed transaction as we call log to journal there between different callbacks. I cant say for sure we dont have bugs there

**Comment by adiholden** [comment]
I got this crash when running your branch executing memtier with eval


F20240204 12:06:55.028923 842494 transaction.cc:538] Check failed: coordinator_state_ == 0u (unsigned char value 4 vs. 0) 
*** Check failure stack trace: ***
F20240204 12:06:55.029067 842480 transaction.cc:538] Check failed: coordinator_state_ == 0u (unsigned char value 4 vs. 0) F20240204 12:06:55.029219 842487 transaction.cc:538] Check failed: coordinator_state_ == 0u (unsigned char value 4 vs. 0) 
*** Check failure stack trace: ***
F20240204 12:06:55.029067 842480 transaction.cc:538] Check failed: coordinator_state_ == 0u (unsigned char value 4 vs. 0) F20240204 12:06:55.029219 842487 transaction.cc:538] Check failed: coordinator_state_ == 0u (unsigned char value 4 vs. 0) 
*** Check failure stack trace: ***
    @     0x55d3f55c9781  google::LogMessage::Fail()
    @     0x55d3f55c9781  google::LogMessage::Fail()
    @     0x55d3f55c9781  google::LogMessage::Fail()
    @     0x55d3f55c96c7  google::LogMessage::SendToLog()
    @     0x55d3f55c96c7  google::LogMessage::SendToLog()
    @     0x55d3f55c96c7  google::LogMessage::SendToLog()
    @     0x55d3f55c8e9c  google::LogMessage::Flush()
    @     0x55d3f55c8e9c  google::LogMessage::Flush()
    @     0x55d3f55ccd14  google::LogMessageFatal::~LogMessageFatal()
    @     0x55d3f55c8e9c  google::LogMessage::Flush()
    @     0x55d3f55ccd14  google::LogMessageFatal::~LogMessageFatal()
    @     0x55d3f55ccd14  google::LogMessageFatal::~LogMessageFatal()
    @     0x55d3f4a145f7  dfly::Transaction::MultiSwitchCmd()
    @     0x55d3f4a145f7  dfly::Transaction::MultiSwitchCmd()
    @     0x55d3f4a145f7  dfly::Transaction::MultiSwitchCmd()
    @     0x55d3f3cf996d  dfly::Service::DispatchCommand()
    @     0x55d3f3d0568f  dfly::Service::CallFromScript()
    @     0x55d3f3d08865  _ZZN4dfly7Service12EvalInternalEN4absl12lts_202308024SpanINS3_IcEEEERKNS0_8EvalArgsEPNS_11InterpreterEPNS_17ConnectionContextEENKUlNS9_8CallArgsEE0_clESD_
    @     0x55d3f3d52cf2  _ZSt13__invoke_implIvRZN4dfly7Service12EvalInternalEN4absl12lts_202308024SpanINS4_IcEEEERKNS1_8EvalArgsEPNS0_11InterpreterEPNS0_17ConnectionContextEEUlNSA_8CallArgsEE0_JSE_EET_St14__invoke_otherOT0_DpOT1_
    @     0x55d3f3d46c15  _ZSt10__invoke_rIvRZN4dfly7Service12EvalInternalEN4absl12lts_202308024SpanINS4_IcEEEERKNS1_8EvalArgsEPNS0_11InterpreterEPNS0_17ConnectionContextEEUlNSA_8CallArgsEE0_JSE_EENSt9enable_ifIX16is_invocable_r_vIT_T0_DpT1_EESI_E4typeEOSJ_DpOSK_
    @     0x55d3f3d3e1cc  _ZNSt17_Function_handlerIFvN4dfly11Interpreter8CallArgsEEZNS0_7Service12EvalInternalEN4absl12lts_202308024SpanINS7_IcEEEERKNS4_8EvalArgsEPS1_PNS0_17ConnectionContextEEUlS2_E0_E9_M_invokeERKSt9_Any_dataOS2_
    @     0x55d3f3cf996d  dfly::Service::DispatchCommand()
    @     0x55d3f3cf996d  dfly::Service::DispatchCommand()
    @     0x55d3f4b16676  std::function<>::operator()()
    @     0x55d3f3d0568f  dfly::Service::CallFromScript()
    @     0x55d3f4b127ab  dfly::Interpreter::RedisGenericCommand()
    @     0x55d3f3d08865  _ZZN4dfly7Service12EvalInternalEN4absl12lts_202308024SpanINS3_IcEEEERKNS0_8EvalArgsEPNS_11InterpreterEPNS_17ConnectionContextEENKUlNS9_8CallArgsEE0_clESD_
    @     0x55d3f3d0568f  dfly::Service::CallFromScript()
    @     0x55d3f3d52cf2  _ZSt13__invoke_implIvRZN4dfly7Service12EvalInternalEN4absl12lts_202308024SpanINS4_IcEEEERKNS1_8EvalArgsEPNS0_11InterpreterEPNS0_17ConnectionContextEEUlNSA_8CallArgsEE0_JSE_EET_St14__invoke_otherOT0_DpOT1_
    @     0x55d3f3d08865  _ZZN4dfly7Service12EvalInternalEN4absl12lts_202308024SpanINS3_IcEEEERKNS0_8EvalArgsEPNS_11InterpreterEPNS_17ConnectionContextEENKUlNS9_8CallArgsEE0_clESD_
    @     0x55d3f3d46c15  _ZSt10__invoke_rIvRZN4dfly7Service12EvalInternalEN4absl12lts_202308024SpanINS4_IcEEEERKNS1_8EvalArgsEPNS0_11InterpreterEPNS0_17ConnectionContextEEUlNSA_8CallArgsEE0_JSE_EENSt9enable_ifIX16is_invocable_r_vIT_T0_DpT1_EESI_E4typeEOSJ_DpOSK_
    @     0x55d3f3d52cf2  _ZSt13__invoke_implIvRZN4dfly7Service12EvalInternalEN4absl12lts_202308024SpanINS4_IcEEEERKNS1_8EvalArgsEPNS0_11InterpreterEPNS0_17ConnectionContextEEUlNSA_8CallArgsEE0_JSE_EET_St14__invoke_otherOT0_DpOT1_
    @     0x55d3f3d3e1cc  _ZNSt17_Function_handlerIFvN4dfly11Interpreter8CallArgsEEZNS0_7Service12EvalInternalEN4absl12lts_202308024SpanINS7_IcEEEERKNS4_8EvalArgsEPS1_PNS0_17ConnectionContextEEUlS2_E0_E9_M_invokeERKSt9_Any_dataOS2_
    @     0x55d3f3d46c15  _ZSt10__invoke_rIvRZN4dfly7Service12EvalInternalEN4absl12lts_202308024SpanINS4_IcEEEERKNS1_8EvalArgsEPNS0_11InterpreterEPNS0_17ConnectionContextEEUlNSA_8CallArgsEE0_JSE_EENSt9enable_ifIX16is_invocable_r_vIT_T0_DpT1_EESI_E4typeEOSJ_DpOSK_
    @     0x55d3f3d3e1cc  _ZNSt17_Function_handlerIFvN4dfly11Interpreter8CallArgsEEZNS0_7Service12EvalInternalEN4absl12lts_202308024SpanINS7_IcEEEERKNS4_8EvalArgsEPS1_PNS0_17ConnectionContextEEUlS2_E0_E9_M_invokeERKSt9_Any_dataOS2_
    @     0x55d3f4b12f35  dfly::Interpreter::RedisCallCommand()
    @     0x55d3f4bc4295  luaD_precall
    @     0x55d3f4b16676  std::function<>::operator()()
    @     0x55d3f4bd2bb4  luaV_execute
    @     0x55d3f4b16676  std::function<>::operator()()
    @     0x55d3f4bc467d  luaD_callnoyield
    @     0x55d3f4b127ab  dfly::Interpreter::RedisGenericCommand()
    @     0x55d3f4b127ab  dfly::Interpreter::RedisGenericCommand()
    @     0x55d3f4bc35e2  luaD_rawrunprotected
    @     0x55d3f4bc4a24  luaD_pcall
    @     0x55d3f4b12f35  dfly::Interpreter::RedisCallCommand()
    @     0x55d3f4b12f35  dfly::Interpreter::RedisCallCommand()
    @     0x55d3f4bc4295  luaD_precall
    @     0x55d3f4bc1050  lua_pcallk
    @     0x55d3f4bc4295  luaD_precall
    @     0x55d3f4bd2bb4  luaV_execute
    @     0x55d3f4b0d143  dfly::Interpreter::RunFunction()
    @     0x55d3f3d08de3  _ZZN4dfly7Service12EvalInternalEN4absl12lts_202308024SpanINS3_IcEEEERKNS0_8EvalArgsEPNS_11InterpreterEPNS_17ConnectionContextEENKUlPNS_11TransactionEPNS_11EngineShardEE1_clESE_SG_
    @     0x55d3f3d46f55  _ZSt13__invoke_implIN6facade8OpStatusERKZN4dfly7Service12EvalInternalEN4absl12lts_202308024SpanINS6_IcEEEERKNS3_8EvalArgsEPNS2_11InterpreterEPNS2_17ConnectionContextEEUlPNS2_11TransactionEPNS2_11EngineShardEE1_JSH_SJ_EET_St14__invoke_otherOT0_DpOT1_
    @     0x55d3f4bc467d  luaD_callnoyield
    @     0x55d3f3d3e401  _ZSt8__invokeIRKZN4dfly7Service12EvalInternalEN4absl12lts_202308024SpanINS4_IcEEEERKNS1_8EvalArgsEPNS0_11InterpreterEPNS0_17ConnectionContextEEUlPNS0_11TransactionEPNS0_11EngineShardEE1_JSF_SH_EENSt15__invoke_resultIT_JDpT0_EE4typeEOSM_DpOSN_
    @     0x55d3f4bd2bb4  luaV_execute
    @     0x55d3f3d35f33  _ZSt6invokeIRKZN4dfly7Service12EvalInternalEN4absl12lts_202308024SpanINS4_IcEEEERKNS1_8EvalArgsEPNS0_11InterpreterEPNS0_17ConnectionContextEEUlPNS0_11TransactionEPNS0_11EngineShardEE1_JSF_SH_EENSt13invoke_resultIT_JDpT0_EE4typeEOSM_DpOSN_
    @     0x55d3f3d2e92d  _ZN4absl12lts_2023080219functional_internal12InvokeObjectIZN4dfly7Service12EvalInternalENS0_4SpanINS5_IcEEEERKNS4_8EvalArgsEPNS3_11InterpreterEPNS3_17ConnectionContextEEUlPNS3_11TransactionEPNS3_11EngineShardEE1_NSF_14RunnableResultEJSG_SI_EEET0_NS1_7VoidPtrEDpNS1_8ForwardTIT1_E4typeE
    @     0x55d3f4bc35e2  luaD_rawrunprotected
    @     0x55d3f4bc467d  luaD_callnoyield
    @     0x55d3f4a4a289  absl::lts_20230802::FunctionRef<>::operator()()
    @     0x55d3f4bc4a24  luaD_pcall
    @     0x55d3f4bc35e2  luaD_rawrunprotected
    @     0x55d3f4bc4a24  luaD_pcall
    @     0x55d3f4a22be2  dfly::Transaction::RunQuickie()
    @     0x55d3f4bc1050  lua_pcallk
    @     0x55d3f4a25aa0  dfly::Transaction::ScheduleUniqueShard()
    @     0x55d3f4a1adad  _ZZN4dfly11Transaction17ScheduleSingleHopEN4absl12lts_2023080211FunctionRefIFNS0_14RunnableResultEPS0_PNS_11EngineShardEEEEENKUlvE_clEv
    @     0x55d3f4b0d143  dfly::Interpreter::RunFunction()
    @     0x55d3f3d08de3  _ZZN4dfly7Service12EvalInternalEN4absl12lts_202308024SpanINS3_IcEEEERKNS0_8EvalArgsEPNS_11InterpreterEPNS_17ConnectionContextEENKUlPNS_11TransactionEPNS_11EngineShardEE1_clESE_SG_
    @     0x55d3f4a3e719  _ZSt13__invoke_implIvRZN4dfly11Transaction17ScheduleSingleHopEN4absl12lts_2023080211FunctionRefIFNS1_14RunnableResultEPS1_PNS0_11EngineShardEEEEEUlvE_JEET_St14__invoke_otherOT0_DpOT1_
    @     0x55d3f3d46f55  _ZSt13__invoke_implIN6facade8OpStatusERKZN4dfly7Service12EvalInternalEN4absl12lts_202308024SpanINS6_IcEEEERKNS3_8EvalArgsEPNS2_11InterpreterEPNS2_17ConnectionContextEEUlPNS2_11TransactionEPNS2_11EngineShardEE1_JSH_SJ_EET_St14__invoke_otherOT0_DpOT1_
    @     0x55d3f4bc1050  lua_pcallk
    @     0x55d3f3d3e401  _ZSt8__invokeIRKZN4dfly7Service12EvalInternalEN4absl12lts_202308024SpanINS4_IcEEEERKNS1_8EvalArgsEPNS0_11InterpreterEPNS0_17ConnectionContextEEUlPNS0_11TransactionEPNS0_11EngineShardEE1_JSF_SH_EENSt15__invoke_resultIT_JDpT0_EE4typeEOSM_DpOSN_
    @     0x55d3f3d35f33  _ZSt6invokeIRKZN4dfly7Service12EvalInternalEN4absl12lts_202308024SpanINS4_IcEEEERKNS1_8EvalArgsEPNS0_11InterpreterEPNS0_17ConnectionContextEEUlPNS0_11TransactionEPNS0_11EngineShardEE1_JSF_SH_EENSt13invoke_resultIT_JDpT0_EE4typeEOSM_DpOSN_
    @     0x55d3f4a3d02b  _ZSt10__invoke_rIvRZN4dfly11Transaction17ScheduleSingleHopEN4absl12lts_2023080211FunctionRefIFNS1_14RunnableResultEPS1_PNS0_11EngineShardEEEEEUlvE_JEENSt9enable_ifIX16is_invocable_r_vIT_T0_DpT1_EESE_E4typeEOSF_DpOSG_
    @     0x55d3f3d2e92d  _ZN4absl12lts_2023080219functional_internal12InvokeObjectIZN4dfly7Service12EvalInternalENS0_4SpanINS5_IcEEEERKNS4_8EvalArgsEPNS3_11InterpreterEPNS3_17ConnectionContextEEUlPNS3_11TransactionEPNS3_11EngineShardEE1_NSF_14RunnableResultEJSG_SI_EEET0_NS1_7VoidPtrEDpNS1_8ForwardTIT1_E4typeE
*** SIGABRT received at time=1707041215 on cpu 1 ***
PC: @     0x7fe2107cd9fc  (unknown)  pthread_kill
    @     0x55d3f4b0d143  dfly::Interpreter::RunFunction()
    @     0x55d3f3d08de3  _ZZN4dfly7Service12EvalInternalEN4absl12lts_202308024SpanINS3_IcEEEERKNS0_8EvalArgsEPNS_11InterpreterEPNS_17ConnectionContextEENKUlPNS_11TransactionEPNS_11EngineShardEE1_clESE_SG_
    @     0x55d3f3d46f55  _ZSt13__invoke_implIN6facade8OpStatusERKZN4dfly7Service12EvalInternalEN4absl12lts_202308024SpanINS6_IcEEEERKNS3_8EvalArgsEPNS2_11InterpreterEPNS2_17ConnectionContextEEUlPNS2_11TransactionEPNS2_11EngineShardEE1_JSH_SJ_EET_St14__invoke_otherOT0_DpOT1_
    @     0x55d3f3d3e401  _ZSt8__invokeIRKZN4dfly7Service12EvalInternalEN4absl12lts_202308024SpanINS4_IcEEEERKNS1_8EvalArgsEPNS0_11InterpreterEPNS0_17ConnectionContextEEUlPNS0_11TransactionEPNS0_11EngineShardEE1_JSF_SH_EENSt15__invoke_resultIT_JDpT0_EE4typeEOSM_DpOSN_
    @     0x55d3f3d35f33  _ZSt6invokeIRKZN4dfly7Service12EvalInternalEN4absl12lts_202308024SpanINS4_IcEEEERKNS1_8EvalArgsEPNS0_11InterpreterEPNS0_17ConnectionContextEEUlPNS0_11TransactionEPNS0_11EngineShardEE1_JSF_SH_EENSt13invoke_resultIT_JDpT0_EE4typeEOSM_DpOSN_
    @     0x55d3f3d2e92d  _ZN4absl12lts_2023080219functional_internal12InvokeObjectIZN4dfly7Service12EvalInternalENS0_4SpanINS5_IcEEEERKNS4_8EvalArgsEPNS3_11InterpreterEPNS3_17ConnectionContextEEUlPNS3_11TransactionEPNS3_11EngineShardEE1_NSF_14RunnableResultEJSG_SI_EEET0_NS1_7VoidPtrEDpNS1_8ForwardTIT1_E4typeE
    @     0x55d3f4a4a289  absl::lts_20230802::FunctionRef<>::operator()()
    @     0x55d3f564e8ef         64  absl::lts_20230802::WriteFailureInfo()
    @     0x55d3f4a4a289  absl::lts_20230802::FunctionRef<>::operator()()
    @     0x55d3f564eb49         96  absl::lts_20230802::AbslFailureSignalHandler()
    @     0x7fe210779520  (unknown)  (unknown)
Aborted (core dumped)

**Comment by dranikpg** [comment]
1. Fixed DCHECK error, it was harmless, but let's better keep the stronger dcheck
2. Removed preemption

**Review by adiholden — DISMISSED** [review]

**Review by romange — DISMISSED** [review]

File: src/server/transaction.cc [hunk 2]
```diff
@@ -1503,9 +1503,8 @@ void Transaction::FinishLogJournalOnShard(EngineShard* shard, uint32_t shard_cnt
 
 void Transaction::CancelBlocking(std::function<OpStatus(ArgSlice)> status_cb) {
   // We're on the owning thread of this transaction, so we can safely access it's data below.
-  // We still need to claim the blocking barrier, but as this function is often called blindly, we
-  // want to check first if it makes sense to even proceed.
-  if (blocking_barrier_.IsClaimed())
+  // First, check if it makes sense to proceed.
+  if (blocking_barrier_.IsClaimed() || (cid_->opt_mask() & CO::BLOCKING) == 0)
```
Comment (adiholden) [inline src/server/transaction.cc:1507]: I now see this error
/home/adi/dragonfly/src/server/transaction.cc:1507:55: runtime error: member access within null pointer of type 'const struct CommandId'
/home/adi/dragonfly/src/server/transaction.cc:1507:55: runtime error: member call on null pointer of type 'const struct CommandId'

when runnning memtier with eval

**Comment by dranikpg** [comment]
Making stub transactions accessible from the outside is really not a good decision but it's what we have

**Review by adiholden — APPROVED** [review]

**Pull request merged**